File: 12-BinarySearch/BinarySearch.php
<?php

/**
 * Search number from an integer list by applying Binary Search Algorithm.
 *
 * @param  int   $needle   The number to search for
 * @param  array $haystack The list to search in
 * @return int   Position of needle in list if found, otherwise -1.
*/
function binarySearch(int $needle, array $haystack): int
{
    $low = 0;
    $high = count($haystack) - 1;

    while ($low <= $high) {
        $mid = intval(($low + $high) / 2);
        $guess = $haystack[$mid];
        if ($guess == $needle) {
            return $mid;
        } elseif ($guess < $needle) {
            $low = $mid + 1;
        } elseif ($guess > $needle) {
            $high = $mid - 1;
        }
    }

    return -1;
}

$num = rtrim(fgets(STDIN));
$stackInput = explode(' ', rtrim(fgets(STDIN)));
$haystack = array_slice($stackInput, 0, $num);
$needle = rtrim(fgets(STDIN));

$result = binarySearch($needle, $haystack);
echo 'result: ';
echo ($result != -1) ? $result : 'none';
echo "\n";

File: 12-BinarySearch/binary_search.cpp
#include <iostream>
#include <vector>

using namespace std;

/**
 * Search number from an integer list by applying Binary Search Algorithm.
 * The list is required sorted.
 *
 * @param  needle   The number to search for
 * @param  haystack The sorted list to search in
 * @return Position of needle in list if found, otherwise -1.
*/
int binarySearch(int needle, vector<int> haystack) {
    int low = 0;
    int high = haystack.size() - 1;
    int mid, guess;
    
    while (low <= high) {
        mid = (low + high) /2;
        guess = haystack[mid];
        if (guess == needle) {
            return mid;
        } else if (guess < needle) {
            low = mid + 1;
        } else if (guess > needle) {
            high = mid -1;
        }
    }
    
    return -1;
}

int main() {
    int num, needle, input, result;
    vector<int> haystack;
    
    cin >> num;
    for (int i = 0; i < num; ++i) {
        cin >> input;
        haystack.push_back(input);
    }
    cin >> needle;
    
    result = binarySearch(needle, haystack);
    cout << "result: ";
    if (result != -1) {
        cout << result;
    } else {
        cout << "none";
    }
    cout << "\n";
    
    return 0;
}